corehq/apps/registration/utils.py

```diff
@@ -92,27 +93,35 @@ def safe_unsubscribe_user_from_mailchimp_list(user, list_id, email=None):
         logging.error(e.message)
 
 
-def handle_changed_mailchimp_email(user, old_email, new_email, list_id):
-    def is_user_subscribed_with_email(couch_user):
-        return (couch_user.subscribed_to_commcare_users
-                and couch_user.email == old_email)
-    users_subscribed_with_old_email = [
-        other_user
-        for other_user in CouchUser.all()
-        if is_user_subscribed_with_email(other_user)
-    ]
+def handle_changed_mailchimp_email(user, old_email, new_email):
+    """
+        Checks whether there are other users with old_email who are also subscribed to any mailchimp lists.
+        If not, it safely unsubscribes that email from mailchimp. Then, adds new_email to mailchimp.
+    """
+    users_with_old_email = User.objects.filter(email=old_email).values_list('username', flat=True)
+
+    users_subscribed_with_old_email = []
+    couch_users = CouchUser.get_db().view('users/by_username',
+                                          keys=list(users_with_old_email),
+                                          include_docs=True,
+                                          reduce=False,
+                                          ).all()
+    for couch_user in couch_users:
+        if couch_user['doc']['subscribed_to_commcare_users'] or not couch_user['doc']['email_opt_out']:
```

Comment:
for the first question, you can't necessarily make that assumption if you're working with the raw JSON coming out of the database. If you work with the wrapped objects (so `CouchUser.view` instead of `CouchUser.get_db().view` which will automatically unpack the doc from the result and wrap it with the class) then `jsonobject` will handle the wrapping for you and let you reference `user.subscribed_to_commcare_users`). That said it's a performance tradeoff since wrapping takes additional work. In this case since we expect the number of docs to be quite small, wrapping is probably fine. 

Also I like list comprehensions, but I don't think there's an explicit style guide around that.